### PR TITLE
API-3182 JSON Schema Errors

### PR DESCRIPTION
**Pull request by cilestin**
## Description of change
Changed NOD error handling to follow [JSON API 1.0 ErrorObject](https://jsonapi.org/format/1.0/#error-objects) formatting.

Extracted json schema errors out into the Common::Exceptions::DetailedSchemaErrors class which currently supports:
* Invalid enum values (adds `meta` details of the allowed values)
* Pattern mismatches (adds `meta` details of the expected regex)
* Missing required fields (adds `meta` details of the missing fields)
* Data type mismatch (e.g. expected boolean, got string)
* Maximum & minimum lengths (adds `meta` details of the min/max lengths)
* Range outliers (adds `meta` details of the min/max allowed for the range)
* Extra data provided that is not in the schema
* Invalid Arrays of items (adds `meta` details about size constraints & uniqueness)
* Error titles & details are translated

Custom validations & error messages are expected to manually add JSON API 1.0 ErrorObject formatted hashes to their model errors. They key of the error should be the pointer path (`data_pointer`) of the field that caused the error. See `NoticeOfDisagreement#validate_hearing_type_selection` for an example.

## Original issue(s)
https://vajira.max.gov/browse/API-3182

#### Example JSON response from multiple types of schema & custom validations:
```json
{
  "errors": [
    {
      "title": "Invalid option",
      "detail": "'invalidType' is not an available option",
      "code": "151",
      "source": {
        "pointer": "/data/type"
      },
      "status": "422",
      "meta": {
        "available_options": [
          "noticeOfDisagreement"
        ]
      }
    },
    {
      "title": "Missing required fields",
      "detail": "One or more expected fields were not found",
      "code": "155",
      "source": {
        "pointer": "/data/attributes"
      },
      "status": "422",
      "meta": {
        "missing_fields": [
          "socOptIn"
        ]
      }
    },
    {
      "title": "Invalid data type",
      "detail": "Expected boolean data",
      "code": "150",
      "source": {
        "pointer": "/data/attributes/veteran/homeless"
      },
      "status": "422"
    },
    {
      "title": "Missing required fields",
      "detail": "One or more expected fields were not found",
      "code": "155",
      "source": {
        "pointer": "/data/attributes/veteran/address"
      },
      "status": "422",
      "meta": {
        "missing_fields": [
          "addressLine1"
        ]
      }
    },
    {
      "title": "Invalid format",
      "detail": "'30012xxx' did not match the defined format",
      "code": "153",
      "source": {
        "pointer": "/data/attributes/veteran/address/zipCode5"
      },
      "status": "422",
      "meta": {
        "regex": "^[0-9]{5}$"
      }
    },
    {
      "title": "Schema mismatch",
      "detail": "Unknown data provided",
      "code": "156",
      "source": {
        "pointer": "/data/attributes/claimant"
      },
      "status": "422"
    },
    {
      "title": "Invalid array",
      "detail": "The 2 items provided did not match the definition",
      "code": "157",
      "source": {
        "pointer": "/favoriteFruits"
      },
      "status": "422",
      "meta": {
        "received_size": 2,
        "received_unique_items": true,
        "max_items": 5,
        "min_items": 3,
        "unique_items": true
      }
    }
  ]
}
```

### CODE REVIEW

**Comment by cilestin** [comment]
TODO: Expand controller tests to check error messages directly

File: modules/appeals_api/app/controllers/appeals_api/v1/decision_reviews/notice_of_disagreements_controller.rb [hunk 1]
```diff
@@ -1,46 +1,37 @@
...
-      render json: {
-        data: {
-          type: 'noticeOfDisagreementValidation',
-          attributes: { status: 'valid' }
-        }
-      }
-    else
-      render_model_errors
-    end
+    render json: validation_success
```
Comment (yakschuss) [inline modules/appeals_api/app/controllers/appeals_api/v1/decision_reviews/notice_of_disagreements_controller.rb:34]: Are we not concerned with rendering validation errors from `/validate` after this?

Comment (yakschuss) [reply modules/appeals_api/app/controllers/appeals_api/v1/decision_reviews/notice_of_disagreements_controller.rb:34]: Nevermind, I see the callback that renders the json directly.

File: modules/appeals_api/app/controllers/appeals_api/v1/decision_reviews/notice_of_disagreements_controller.rb [hunk 2]
```diff
@@ -51,6 +42,32 @@ def schema
 
   private
 
+  def validate_json_schema
+    validate_json_schema_for_headers
+    validate_json_schema_for_body
+  rescue JsonSchema::JsonApiMissingAttribute => e
```
Comment (yakschuss) [inline modules/appeals_api/app/controllers/appeals_api/v1/decision_reviews/notice_of_disagreements_controller.rb:48]: You wanted me to point out this raise while we talked!

**Comment by ElleeB** [comment]
Just took a glance - looks amazing! I suggest waiting to merge until my 'USE ON FILE' for address, phone, and email PR is merged, since there are some bits here that will be obsoleted.

File: modules/appeals_api/app/controllers/appeals_api/v1/decision_reviews/notice_of_disagreements_controller.rb [hunk 3]
```diff
@@ -45,16 +46,16 @@ def schema
...
     validate_json_schema_for_body
-  rescue JsonSchema::JsonApiMissingAttribute => e
-    render json: e.to_json_api, status: e.code
+  rescue SCHEMA_ERROR_TYPE => e
+    render json: { errors: e.errors }, status: 422
   end
 
   def validate_json_schema_for_headers
-    AppealsApi::FormSchemas.new.validate!("#{self.class::FORM_NUMBER}_HEADERS", headers)
+    AppealsApi::FormSchemas.new(SCHEMA_ERROR_TYPE).validate!("#{FORM_NUMBER}_HEADERS", headers)
```
Comment (yakschuss) [inline modules/appeals_api/app/controllers/appeals_api/v1/decision_reviews/notice_of_disagreements_controller.rb:54]: Something about the way this is reading is yelling at me. Why are we validating against an error type?

File: lib/common/exceptions/detailed_schema_errors.rb [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    class DetailedSchemaErrors < BaseError
+      # Expects array of JSONSchemer errors
+      def initialize(raw_schema_errors)
+        @raw_errors = raw_schema_errors
+        raise TypeError, 'the resource provided has no errors' if raw_schema_errors.blank?
+      end
+
+      def errors
+        @raw_errors.map do |raw_error|
+          type = raw_error['type'].downcase
```
Comment (yakschuss) [inline lib/common/exceptions/detailed_schema_errors.rb:17]: Is the list of possible types documented?

**Review by yakschuss — APPROVED** [review]
I like this! I'm hoping that we'll be able to move away from `JsonSchema::JsonApiMissingAttribute` and reach a standard quickly, but the way this was added without any breaking changes was well done. 🥳

File: lib/common/exceptions/detailed_schema_errors.rb (hunk 4 above)
Comment (cilestin) [reply lib/common/exceptions/detailed_schema_errors.rb:17]: I think the fallback I just added should be self-documenting enough show what's supported, but I can add in a constant to check instead if you think that'd make things clearer.

**Pull request merged**